src/libsystemd-network/dhcp-network.c

```diff
@@ -186,16 +186,18 @@ int dhcp_network_bind_udp_socket(int ifindex, be32_t address, uint16_t port, int
                         return r;
         }
 
+        r = setsockopt_int(s, SOL_SOCKET, SO_BROADCAST, true);
+        if (r < 0)
+                return r;
+
         if (address == INADDR_ANY) {
+                /* IP_PKTINFO filter should not be applied when packets are
+                   allowed to enter/leave through the interface other than
+                   DHCP server sits on(BindToInterface option). */
                 r = setsockopt_int(s, IPPROTO_IP, IP_PKTINFO, true);
                 if (r < 0)
                         return r;
-
-                r = setsockopt_int(s, SOL_SOCKET, SO_BROADCAST, true);
-                if (r < 0)
-                        return r;
-
-        } else {
+        } else if (address != INADDR_BROADCAST) {
                 r = setsockopt_int(s, IPPROTO_IP, IP_FREEBIND, true);
```

Comment:
Originally, this option is for dhcp client. Please change the condition to `if (port != DHCP_PORT_SERVER)`.